pkg/status/singletonstatus.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// Should be wr-locked
+type singletonState struct {
+	wObjSync map[util.ObjectIdentifier]bool
+}
+
+func (ss *singletonState) Get(id util.ObjectIdentifier) bool {
+	return ss.wObjSync[id]
+}
+
+func (ss *singletonState) Set(id util.ObjectIdentifier) {
+	ss.wObjSync[id] = true
+}
+
+func (ss *singletonState) Unset(id util.ObjectIdentifier) {
+	ss.wObjSync[id] = false
+}
+
+// Should it be a field of Contoller?
+var theSingletonState singletonState = singletonState{
+	wObjSync: map[util.ObjectIdentifier]bool{},
+}
+
+func (c *Controller) buildSingletonStateAndOptionallyReconcile(ctx context.Context, reconcile bool) error {
+	logger := klog.FromContext(ctx)
+	logger.Info("Initializing the desired state for singleton statuses")
+
+	allBdgs, err := c.wdsKsClient.ControlV1alpha1().Bindings().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, bdg := range allBdgs.Items {
+		logger.V(2).Info("Got Binding", "name", bdg.Name)
+		list := bdg.Spec.Workload.NamespaceScope // Should also do the same thing for cluster scoped
+		for _, item := range list {
+			nsWObjRef := item.NamespaceScopeDownsyncObject
+			lister, found := c.listers.Get(metav1GVR2schemaGVR(nsWObjRef.GroupVersionResource))
+			if !found {
+				return fmt.Errorf("could not get lister for gvr: %s", nsWObjRef.GroupVersionResource)
+			}
+			nsWObj, err := lister.ByNamespace(nsWObjRef.Namespace).Get(nsWObjRef.Name)
+			if err != nil {
+				return err
+			}
+			labels := nsWObj.(metav1.Object).GetLabels()
+			if v, ok := labels[util.BindingPolicyLabelSingletonStatusKey]; ok {
+				logger.V(2).Info("Found namespace scoped singleton workload object", "name", nsWObjRef.Name)
+				wObjIdentifier := util.ObjectIdentifier{
+					GVK:        schema.GroupVersionKind{Group: nsWObjRef.Group, Version: nsWObjRef.Version, Kind: nsWObj.GetObjectKind().GroupVersionKind().Kind},
+					Resource:   nsWObjRef.Resource,
+					ObjectName: cache.NewObjectName(nsWObjRef.Namespace, nsWObjRef.Name),
+				}
+				if v == util.BindingPolicyLabelSingletonStatusValueSet && len(bdg.Spec.Destinations) > 0 {
```

Comment:
Shouldn't this be `== 1` rather than `> 0` ?

Reply:
I think it should be `>0`, which says this Binding object asks some WEC to be candidates so that the overall set of candidates (after checking all Binding objects) is for sure not empty, thus the status should be synced.